src/file_request_handler.cc

```diff
@@ -65,9 +66,11 @@ FileRequestHandler::FileRequestHandler(std::shared_ptr<ConfigManager> config,
 
 void FileRequestHandler::getInfo(const char* filename, UpnpFileInfo* info)
 {
-    Headers headers;
     log_debug("start");
 
+    auto quirks = std::make_unique<Quirks>(config);
```

Comment:
I see you have sometimes `shared_ptr` for `Quirks`, `unique_ptr` here. I wonder if you really need to create one on heap all the time?

I can imagine if you had more generic name like "Request PostProcessor" with virtual interface which is injected for unit testing... But for example in this particular function it could be just on stack.

I also see through the code base that you toss `shared_ptrs` around quite much and this makes me really think "Is this possible that function does need to potentially take an ownership of the object or is it just "becase shared ptrs are everywhere"?

I think even Cpp Guidelines (for modern C++) encourage use of raw pointers as a "borrowed" object (that is if you don't have `delete` in your code you don't need to think twice whether to deallocate object or not).

I'm asking to understand and deduce coding style here a bit better :)

Reply:
its always a question of speed vs safety. Stack allocation is super fast. My first goal is to not leak any memory and to use an easy to use concept. Whenever speed is not critical do not loose time with optimizations. Focus on hot spots first.

If you have pointers allover in your code how to you know who is responsible for deleting it again? It gets quiet quite messy, specially if more than one person is coding in the same base.

Also consider this: e.g. std::string:
The string object itself is stored on the stack but it points to memory that is on the heap. Maybe some parts is optimized away by the compiler.

Reply:
@pamapa 

I totally agree with your first paragraph.

However I'm most worried not about efficiency but rather semantics and expressiveness: When I (as someone not familiar with the code, but just applying a common sense) read smart pointer I assume that ownership of the object can be transferred somewhere.

* Passing `shared_ptr` by const ref means it can be potentially stored somewhere
* Passing `unique_ptr` by ref means it *could be* stored (maybe under some circumstances).
* Passing by value or `&&` means "sink": it will be stored, but the caller doesn't need it.

The choice `unique_ptr` vs `shared_ptr` has also significant impact on thread-safety requirements on type: in case of `unique_ptr` one can assume "external synchronization" by owner (i.e. mutex or just by design). in case of `shared_ptr` most probably class needs to be thread-safe internally (or have only `const` methods which is an another way to express "thread safe") since it's hard to synchronize access for arbitrary number of owners (since anyone can pass it further to other objects).

That's said I got confused by mixed usage of `unique_ptr` and `shared_ptr` as handlers for a single class. I just can't easily tell guarantee that class must promise.
 Especially confusing looked local `unique_ptr` without passing it somewhere neither using any polymorphic behavior (In this case returning from a factory is a passing of ownership, at least once).

Most of this is based on C++ Core Guidelines, which also recommend usage of *raw* pointers. I really though: Did Bjarne got dementia? when I first read it, but after some thinking it makes perfect sense: If you ban:
1. Explicit calls to `new` and `delete` (which you do)
2. "Adopting" raw pointer into a smart pointer (i.e. allow only `make_*` functions to create objects) - which you also do.

There is just no question "Who needs to delete?". It's simple: no one! (And given above bans you just can't). Essentially raw pointers become just a reference which can be a null or rebound (where you need such behavior). 

I can take that this code base is insanely old and was full of raw pointers and manual memory management so it was plain dangerous to allow raw pointers, but now it seems ok, so maybe it's time to allow this.

Would be really much easier if most projects can just point at C++ Core Guidelines (well, maybe with some exceptions as usual).